### ISSUE TITLE
Can you retreive the result of a JS callback function called by a host object?

### ISSUE THREAD

**Issue by G33kDude** [MicrosoftEdge/WebView2Feedback#515]
I've been experimenting with the AddHostObjectToScript functionality, and am most intrigued by the ability to pass callback functions back and forth between the host application and the WebView2 instance. I've had success in invoking a callback function as defined in JavaScript (such as showing a page alert), but it is not clear to me how to retrieve the result of the callback function after it has finished its asynchronous execution.

Here is some mock JavaScript and C++ pseudo code which may give a bit more insight into the situation. I'm developing my application in another language (not C++) so I may have made some error in the implementation below, but I think it should get the point across.

```js
window.chrome.webview.hostObjects.hostObject.callback_test( () => 42 );
```

```c++
HRESULT callback_test(IDispatch* callbackParameter)
{
    VARIANT result;
    result.vt = 0;
    callbackParameter->Invoke(DISPID_VALUE, IID_NULL, LOCALE_USER_DEFAULT, DISPATCH_METHOD, 0, &result, 0, 0);
    cout << result.vt; // 0
}
```

It appears from my tests that the IDispatch interface for the callback function does not alter the data pointed to by the `pVarResult` parameter. My hope was that it would return a new IDispatch object representing a promise, where I could invoke a `then` method and give it a host object to be called once the result was available.

I suppose I could rewrite the JavaScript to explicitly invoke a host object instead of returning a value, but that seems non-ideal. Is there there a way to retrieve the return value, or, if not, is there planned to be a way to retrieve it?

**Comment by ysc3839** [comment]
Personally I suggest not using host object. Instead, use a helper JavaScript to do the communication.
I modified the JavaScript [code](https://github.com/marcuswestin/WebViewJavascriptBridge/blob/master/WebViewJavascriptBridge/WebViewJavascriptBridge_JS.m) of [WebViewJavascriptBridge](https://github.com/marcuswestin/WebViewJavascriptBridge) to make it works with Edge WebView2 in my project.
If you want, I can make a standalone project.

**Comment by champnic** [comment]
I don't think this is currently supported, so I added a scenario to our backlog. Thanks!

**Comment by rshunter1982** [comment]
@ysc3839 , I'm trying to use WebView2 to get a JWT after login. I'm currently grabbing the html content, is there an easier way?

**Comment by jaredkrinke** [comment]
This issue makes it clear that you can't obtain the *result* of invoking a JavaScript callback set on a native host object from native code via `Invoke`, but is it at least possible to know when the callback has completed?

For what it's worth, it's possible to hack around this limitation using `ExecuteScript`:

* Given a native host object with a property named `OnSomethingHappened` that holds a JavaScript callback
* If that host object is exposed as `chrome.webview.hostObjects.sync.someHostObject` (via `AddHostObjectToScript`)
* You can call `ExecuteScript(L"chrome.webview.hostObjects.sync.someHostObject.OnSomethingHappened()", ...)` and supply `ICoreWebView2ExecuteScriptCompletedHandler` to get the result (as a JSON-ified string)

**Comment by xiaoyaocode163** [comment]
js get result from vb6(host exe) 
vb6 edge webview2 demo by IUnknown，without rc6.dll-VBForums

https://www.vbforums.com/showthread.php?900231-vb6-edge-webview2-demo-by-IUnknown%26%2365292%3Bwithout-rc6-dll
